File: leetcode/tree/1214_Two_Sum_BSTs/codes/inorder.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
class Solution:
    def inorder(self, root, res):
        if not root: return
        self.inorder(root.left, res)
        res.append(root.val)
        self.inorder(root.right, res)

    def twoSumBSTs(self, root1: Optional[TreeNode], root2: Optional[TreeNode], target: int) -> bool:
        arr1, arr2 = [], []
        self.inorder(root1, arr1)
        self.inorder(root2, arr2)
        
        i, j = 0, len(arr2) - 1
        while i < len(arr1) and j >= 0:
            s = arr1[i] + arr2[j]
            if s == target:
                return True
            elif s < target:
                i += 1
            else:
                j -= 1
        return False

File: leetcode/tree/1214_Two_Sum_BSTs/codes/inorder.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};

void inorder(struct TreeNode* root, int* arr, int* idx) {
    if (!root) return;
    inorder(root -> left, arr, idx);
    arr[(*idx)++] = root -> val;
    inorder(root -> right, arr, idx);
}

bool twoSumBSTs(struct TreeNode* root1, struct TreeNode* root2, int target) {
    int arr1[5000], arr2[5000], n1 = 0, n2 = 0;
    inorder(root1, arr1, &n1);
    inorder(root2, arr2, &n2);

    int i = 0, j = n2 - 1;
    while (i < n1 && j >= 0) {
        int sum = arr1[i] + arr2[j];
        if (sum == target) return true;
        else if (sum < target) i++;
        else j--;
    }
    return false;
}
